### ISSUE TITLE
BUG: 

### ISSUE THREAD

**Issue by jowens** [pandas-dev/pandas#58461]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
#!/usr/bin/env python3

import pandas as pd
import re
import requests
from io import StringIO

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)
pd.set_option("display.width", 500)

html = requests.get(
    "https://en.wikipedia.org/wiki/Template:AMD_Radeon_Pro_V_series"
).text

df = pd.read_html(
    StringIO(html),
    match=re.compile("Radeon Pro V620"),
)
print(df)
```

### Issue Description

The bottom right of the ingested table puts entries in the wrong columns near the right side for the last two rows. I did some checking of the HTML source and even though it's got some complex rowspan and colspan directives, it appears to be properly constructed.

<img width="2198" alt="Cursor_and___Documents_working_owensgroup_proj_gpustats__jowens_piecaken" src="https://github.com/pandas-dev/pandas/assets/976970/dfecc613-b159-41fe-8326-14980f0e3239">
<img width="1342" alt="Template_AMD_Radeon_Pro_V_series_-_Wikipedia" src="https://github.com/pandas-dev/pandas/assets/976970/e7223947-056d-4f5e-a5f8-e8115aba3d67">

I acknowledge that I'm using a slightly older pandas than is installed, but I looked through recent issues on and checkins to `read_html` and I don't believe this is fixed/reported.

### Expected Behavior

I expect the column called "Memory / L3 Cache" to only be populated in the last row.

I expect the two power entries in the last two rows to be placed in the "TDP" column.

Most of the right side of the bottom two rows is misplaced.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : bdc79c146c2e32f2cab629be240f01658cfb6cc2
python                : 3.12.3.final.0
python-bits           : 64
OS                    : Darwin
OS-release            : 23.4.0
Version               : Darwin Kernel Version 23.4.0: Fri Mar 15 00:10:42 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T6000
machine               : arm64
processor             : arm
byteorder             : little
LC_ALL                : None
LANG                  : en_US.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 2.2.1
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.8.2
setuptools            : 69.5.1
pip                   : 24.0
Cython                : 3.0.10
pytest                : 7.4.3
hypothesis            : None
sphinx                : 7.2.6
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : 5.2.1
html5lib              : 1.1
pymysql               : None
psycopg2              : None
jinja2                : 3.1.3
IPython               : 8.23.0
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.2
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : None
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
pyarrow               : 13.0.0.dev0+gb7d2f7ffc.d20240415
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : 1.11.4
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None

</details>

**Comment by samukweku** [comment]
@jowens   If you use another tool for the extraction other than pandas, do you get a different result?

**Comment by attack68** [comment]
I didn't implement any of this, and haven't checked the implementation but my guess is going to be that;

a) reading a grid based table is straight forward.
b) accounting for `rowspan` or `colspan` separately is an extension to a) which is not too difficult.
c) accounting for simulataneous cross-over of `rowspan` and `colspan` is really difficult and needs initial passes or very specific structuring. It probably isn't tested.

Can anyone confirm?

**Comment by jowens** [comment]
> If you use another tool for the extraction other than pandas, do you get a different result?

Suggestion for that other tool? I'm happy to try.

**Comment by samukweku** [comment]
@jowens a quick search on google gives this [html-extractor](https://github.com/yuanxu-li/html-table-extractor) - havent used it though (caveat). i asked the earlier question, to see if there is a tool that does it right and we can compare against them. it seems @attack68 has looked into your question more and may have figured out the possible bug?

**Comment by jowens** [comment]
Yeah, that didn't appear to parse correctly. Here's a quick test:

https://gist.github.com/jowens/bd15b42accaa20e9c403af89719a5256

(which just has the table manually in the source code, and then the example in the html_table_extractor docs)

with output

```
$ ./table-extractor.py
[]
[['1', '2'], ['3', '4']]
```

**Comment by jowens** [comment]
Just for posterity, here's the specific Wikipedia revision we're discussing here, in case it gets edited:

https://en.wikipedia.org/w/index.php?title=Template:AMD_Radeon_Pro_V_series&oldid=1220301074

and here's a gist where I extracted everything between `<table>` and `</table>`:

https://gist.github.com/jowens/8e42fa17a5af4bc16284cfab56ef1473

**Comment by jowens** [comment]
html_table_extractor has similar behavior (same errors). Here's a quick test:

https://gist.github.com/jowens/bd15b42accaa20e9c403af89719a5256

(which just has the table manually in the source code).

Here's the last line of the output, which corresponds to what's in the issue description.

```
['Radeon Pro V620(Navi\xa021)[10][11]\n', 'Nov 4, 2021\n', 'RDNA 2TSMC\xa0N7\n', '26.8×109520 mm2\n', '4608:288:128:7272 CU\n', '18252200\n', '525.6633.6\n', '233.6281.6\n', '33,63840,550\n', '16,81920,275\n', '1,0511,267\n', '32\n', '512\n', 'GDDR6256-bit\n', '128 MB\n', '16000\n', 'PCIe\xa04.0×16\n', '—\n', '300\xa0W\n']
```

**Comment by jowens** [comment]
FWIW I just tested with `read_html`'s `flavor="lxml"` and `flavor="bs4"` and they both returned identical results.

**Comment by attack68** [comment]
So is the summary here that all tools you have tested for parsing this table, including pandas, return the same results, **and** that those results are all incorrect.

**Comment by jowens** [comment]
Well, two tools (pandas and html_table_extractor), and those two tools return consistent but incorrect results, where incorrect is compared to how a web browser renders it.

Since these two tools both (appear to) have different code that parses the table's cells / rowspans / colspans, it seems like a possibility that web browsers (I looked at Chrome/Firefox/Safari, each of which [I think] uses a different back end [Chromium/Gecko/WebKit]) might interpret the table differently than these two tools. Web browsers are surely more forgiving of HTML errors.

**Comment by lawcal** [comment]
Ah, I believe the problem is that the original table has 7 nested tables which neither `pandas` nor `html-table-extractor` handles currently. For example, one of the nested tables look like this:

<table><tr><td>2×</td><td>12.5×109<br>495 mm2</td></tr></table>

To scratch my own itch (but mostly because I want more flexibility to interpret the output), I've created [html-table-takeout](https://github.com/lawcal/html-table-takeout) which should handle nested tables correctly on top of rowspan and colspan.

```
from html_table_takeout import parse_html

tables = parse_html('https://en.wikipedia.org/wiki/Template:AMD_Radeon_Pro_V_series')
print(tables[0].to_html())
```

And this is the output HTML. I've added borders to show the nested tables:

![Image](https://github.com/user-attachments/assets/d9b0b0b2-8251-4329-95dd-5d2dee2f9586)

If anybody wants to port the implementation back to Pandas, feel free to copy any parts over!

**Comment by jowens** [comment]
Confirming this is still present on pandas 2.3.3, and adding a network-free minimal reproducer plus a more specific root-cause pointer that builds on @lawcal's "nested tables" diagnosis.

## Still reproduces on current versions

Identical wrong output from all three parser backends, so the bug is in pandas' shared table-construction code (`pandas/io/html.py`), not in a specific backend:

| flavor      | V520 `L3 cache` (expected `—`) | V520 `Memory Size (GB)` (expected `8`) |
| ----------- | ------------------------------ | -------------------------------------- |
| `lxml`      | `'8'`                          | `'512'`                                |
| `html5lib`  | `'8'`                          | `'512'`                                |
| `bs4`       | `'8'`                          | `'512'`                                |

Versions: pandas 2.3.3, lxml 6.0.4, beautifulsoup4 4.14.3, html5lib 1.1, Python 3.14.4, macOS arm64.

## Minimal self-contained reproducer (no network)

```python
import pandas as pd
from io import StringIO

HTML = """<table><tbody>
<tr>
  <th rowspan="2">Model</th><th rowspan="2">Release</th>
  <th rowspan="2">Arch</th><th rowspan="2">Trans</th>
  <th colspan="2">Core</th><th colspan="2">Fillrate</th>
  <th colspan="3">Power</th>
  <th rowspan="2">L3 cache</th>
  <th colspan="4">Memory</th>
  <th rowspan="2">TDP</th><th rowspan="2">Bus</th><th rowspan="2">Out</th>
</tr>
<tr>
  <th>Config</th><th>Clock</th><th>Tex</th><th>Pix</th>
  <th>Half</th><th>Single</th><th>Double</th>
  <th>Size</th><th>BW</th><th>BusType</th><th>MClock</th>
</tr>
<tr>
  <th>R1</th><td>2017</td>
  <td rowspan="2">GCN5</td>
  <td>12.5e9</td><td>4096</td><td>852</td>
  <td>218</td><td>54</td>
  <td>13959</td><td>6979</td><td>436</td>
  <td rowspan="3">-</td>
  <td>16</td><td>484</td><td>HBM2-2048</td>
  <td rowspan="2">1890</td>
  <td>230W</td>
  <td rowspan="2">PCIe3</td>
  <td>4xDP</td>
</tr>
<tr>
  <th>R2</th><td>2018</td>
  <td><table><tbody><tr><td>2x</td><td>NESTED</td></tr></tbody></table></td>
  <td>XX</td><td>852</td>
  <td>XX</td><td>XX</td><td>XX</td><td>XX</td><td>XX</td>
  <td>2x16</td><td>2x484</td><td>HBM2-2x2048</td>
  <td>300W</td><td>1xDP</td>
</tr>
<tr>
  <th>R3</th><td>2020</td><td>RDNA</td><td>?</td>
  <td>2304</td><td>1000</td>
  <td>144</td><td>64</td>
  <td>9216</td><td>4608</td><td>288</td>
  <td>8</td><td>512</td><td>HBM2-2048</td>
  <td>2000</td><td>225W</td><td>PCIe4</td><td>-</td>
</tr>
</tbody></table>"""

print(pd.read_html(StringIO(HTML))[0].to_string())
```

### Expected (R3 row)

```
L3 cache = '-'          Size = '8'             BW      = '512'
BusType  = 'HBM2-2048'  MClock = '2000'        TDP     = '225W'
Bus      = 'PCIe4'      Out  = '-'
```

### Actual (R3 row)

```
L3 cache = '8'          Size = '512'           BW      = 'HBM2-2048'
BusType  = '2000'       MClock = '225W'        TDP     = 'PCIe4'
Bus      = '-'          Out  = NaN
```

Every column from the L3-cache position onward is shifted left by one for R3 (the row *after* the row with the nested table).

## Isolation findings

- Remove the nested `<table>` in R2 → **bug disappears.**
- Keep the nested `<table>` but remove `rowspan="3"` → **bug disappears.**
- Adding more nested tables in R2 (the real table has 7) does not change the result — **one nested table is sufficient.**
- The other `rowspan="2"` cells in R1 are not required.
- Number of columns before the `rowspan="3"` cell does not matter.
- R2 itself correctly receives the `rowspan="3"` inherited value. Only R3 (and any further rows covered by the rowspan) are affected.

## Root-cause pointer

Printing the parsed DataFrame reveals a **phantom row emitted between R2 and R3** carrying the contents of the nested `<table>`:

```
0  R1  2017  GCN5    12.5e9  ...  -    16    484        ...
1  R2  2018  GCN5  2xNESTED  ...  -  2x16  2x484        ...
2  2x NESTED    -       NaN  NaN NaN  NaN   NaN         ...     <-- phantom
3  R3  2020  RDNA       ?   ...  8   512   HBM2-2048    ...     <-- shifted
```

That phantom row appears to consume one of the three slots of the `rowspan="3"`, which is why R3 — the real third target of the rowspan — no longer receives the inherited L3 cell, and every subsequent column ends up shifted left by one.

In other words the bug isn't that rowspan is broken; it's that nested tables inflate the outer table's row count, and the rowspan bookkeeping is not aware of it. A fix likely needs `_expand_elements` / the colspan & rowspan accounting in `pandas/io/html.py` to either (a) skip cells whose children are `<table>` when tallying outer rows, or (b) stop emitting the nested-table contents as extra rows of the outer table.

## Workaround

For anyone landing here: @lawcal's [html-table-takeout](https://github.com/lawcal/html-table-takeout) library handles nested tables correctly and is a near-drop-in replacement for this specific failure mode.

Thanks @claude for working out this comment with me.
